addons/maintenance/models/maintenance.py

```diff
@@ -326,11 +326,13 @@ def write(self, vals):
         # the stage (stage_id) of the Maintenance Request changes.
         if vals and 'kanban_state' not in vals and 'stage_id' in vals:
             vals['kanban_state'] = 'normal'
-        if 'stage_id' in vals and self.maintenance_type == 'preventive' and self.recurring_maintenance and self.env['maintenance.stage'].browse(vals['stage_id']).done:
-            schedule_date = self.schedule_date or fields.Datetime.now()
-            schedule_date += relativedelta(**{f"{self.repeat_unit}s": self.repeat_interval})
-            if self.repeat_type == 'forever' or schedule_date.date() <= self.repeat_until:
-                self.copy({'schedule_date': schedule_date, 'stage_id': self._default_stage().id})
+        now = fields.Datetime.now()
+        for request in self:
+            if 'stage_id' in vals and request.maintenance_type == 'preventive' and request.recurring_maintenance and request.env['maintenance.stage'].browse(vals['stage_id']).done:
+                schedule_date = request.schedule_date or now
+                schedule_date += relativedelta(**{f"{request.repeat_unit}s": request.repeat_interval})
+                if request.repeat_type == 'forever' or schedule_date.date() <= request.repeat_until:
+                    request.copy({'schedule_date': schedule_date, 'stage_id': request._default_stage().id})
```

Comment:
Fix looks good, but I think it will be cleaner/clearer if you pull out the common `if` conditions and use the standard way of using `browse` (use `self.env['model.name']` since it's clearer that it's unrelated to the record you're currently looking at). Also inverting the internal if into a `or` condition is a minor short-circuiting optimization + will allow us to avoid indenting a lot:

```suggestion
        if 'stage_id' in vals and self.env['maintenance.stage'].browse(vals['stage_id']).done:
            for request in self:
                if request.maintenance_type != 'preventive' or not request.recurring_maintenance:
                    continue
                schedule_date = request.schedule_date or now
                schedule_date += relativedelta(**{f"{request.repeat_unit}s": request.repeat_interval})
                if request.repeat_type == 'forever' or schedule_date.date() <= request.repeat_until:
                    request.copy({'schedule_date': schedule_date, 'stage_id': request._default_stage().id})
```